ouroboros-network/test/Test/Ouroboros/Network/PeerSelection.hs

```diff
@@ -156,16 +154,151 @@ isEmptyEnv GovernorMockEnvironment {
       || all (\(t,_) -> targetNumberOfRootPeers  t == 0) targets)
 
 
--- Check that events that are 100 events apart have an adequate time
--- between them, to indicate we're not in a busy livelock situation.
-makesAdequateProgress :: Int -> DiffTime -> [Time] -> Bool
-makesAdequateProgress n adequate ts =
-    go ts (drop n ts)
+-- | As a basic property we run the governor to explore its state space a bit
+-- and check it does not throw any exceptions (assertions such as invariant
+-- violations).
+--
+-- We do /not/ assume freedom from livelock for this property, so we run the
+-- governor for a maximum number of trace events rather than for a fixed
+-- simulated time.
+--
+prop_governor_nofail :: GovernorMockEnvironment -> Bool
+prop_governor_nofail env =
+    let trace = take 5000 .
+                selectPeerSelectionTraceEvents $
+                  runGovernorInMockEnvironment env
+
+     in foldl' (flip seq) True
+          [ assertPeerSelectionState st ()
+          | (_, GovernorDebug (TraceGovernorState _ _ st)) <- trace ]
+
+
+-- | It is relatively easy to write bugs where the governor is stuck in a tight
+-- cycle of continuous activity. Due to the way the I\/O simulator manages
+-- virtual time, these bugs exhibits themselves by infinite trace activity
+-- without time advancing.
+--
+-- It is important to catch these bugs early in the set of tests, since it is
+-- hard to write many of the other more interesting properties if there are
+-- these kinds of livelock bugs. Or to put it another way, the other properties
+-- can be expressed more simply if they can assume within event traces that the
+-- time always advances after some finite number of events.
+--
+prop_governor_nolivelock :: GovernorMockEnvironment -> Property
+prop_governor_nolivelock env =
+    let trace = take 500 .
+                selectGovernorEvents .
+                selectPeerSelectionTraceEvents $
+                  runGovernorInMockEnvironment env
+     in case tooManyEventsBeforeTimeAdvances 1000 trace of
+          Nothing -> property True
+          Just (t, es) ->
+            counterexample
+              ("over 1000 events at time: " ++ show t ++ "\n" ++
+               "first 50 events: " ++ (unlines . map show . take 50 $ es)) $
+            property True
```

Comment:
This property never fail for two reasons:

1. this branch should return `property False`
2. one takes only `500` events with threshold of `1000` events in the same time slot.   I guess, we want to `take 5000` as in other tests.

Reply:
Yup. Thanks